hands-on-exercise/1-native/3-add-game.md

```diff
@@ -0,0 +1,314 @@
+
+Use [a ready-made implementation](https://github.com/batkinson/checkers-go/blob/a09daeb/checkers/checkers.go) with the additional rule that the board is 8x8, is played on black cells, and black plays first. This code will not need adjustments. Copy this rules file into a `rules` folder inside your module. Change its package from `checkers` to `rules`. You can do this by command-line:
+
+```sh
+$ mkdir rules
+$ curl https://raw.githubusercontent.com/batkinson/checkers-go/a09daeb1548dd4cc0145d87c8da3ed2ea33a62e3/checkers/checkers.go | sed 's/package checkers/package rules/' > rules/checkers.go
+```
+
+## A stored game object
+
+With the rules in place, begin with the minimum game information needed to be stored.
+
+### The game object type
+
+You need:
+
+* **Index.** A string, so it can be identified and retrieved from storage.
+* **Black player.** A string, the serialized address.
+* **Red player.** A string, the serialized address.
+* **Board proper.** A string, the board as it is serialized by the _rules_ file.
+* **Player to play next.** A string, specifying whose _turn_ it is.
+
+<HighlightBox type="remember">
+
+When you save strings, it makes it easier to understand what comes straight out of storage, but at the expense of storage space. As an advanced consideration, you could store the same information in binary.
+
+</HighlightBox>
+
+As ever, taking inspiration from `minimal-module-example`, this can be described in `proto/../types.proto` with:
+
+```diff-protobuf [proto/alice/checkers/v1/types.proto]
+    message GenesisState {
+      // params defines all the parameters of the module.
+      Params params = 1 [ (gogoproto.nullable) = false, (amino.dont_omitempty) = true ];
+    }
+
++  message StoredGame {
++    option (amino.name) = "alice/checkers/StoredGame"; 
++
++    string index = 1 ;
++    string board = 2;
++    string turn = 3;
++    string black = 4 [(cosmos_proto.scalar) = "cosmos.AddressString"];
++    string red = 5 [(cosmos_proto.scalar) = "cosmos.AddressString"];
++  }
+```
+
+Compile it:
+
+```sh
+$ make proto-gen
+```
+
+Now that you have the individual stored game structure, you can define how it is kept in storage.
+
+### The storage structure
+
+The way that makes sense is to keep a map of games and rely on the SDK's basic structures to optimize saving and retrieving.
+
+Update your `GenesisState` in `types.proto`:
+
+```diff-protobuf
+    message GenesisState {
+      // params defines all the parameters of the module.
+      Params params = 1 [ (gogoproto.nullable) = false, (amino.dont_omitempty) = true ];
+
++     repeated StoredGame storedGameList = 2 [(gogoproto.nullable) = false, (amino.dont_omitempty) = true];
+    }
+```
+
+And recompile.
+
+### Add validation
+
+You can consider adding functions on the `StoredGame`, for instance a `Validate` one. Start with a new `errors.go` file.
```

Comment:
Where should the `errors.go` file be located?

Reply:
In the root folder. I hesitate writing `/errors.go`.

Eventually, there will be a repo with the files, and "View source" would link to it in the repo so people will answer this question by clicking on the link.

Reply:
You could say "located in the root folder of the XYZ repo"